doc/xferlog.5.in
.\"
.\" Copyright (c) 1999-2003 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id: xferlog.5.in,v 1.4 2009/04/19 10:35:36 wmaton Exp $
.\"
.TH XFERLOG 5
.SH NAME
xferlog \- FTP server transfer logfile
.SH DESCRIPTION
.LP
The
.B xferlog
file contains transfer logging information from the FTP server,
.BR ftpd (8).
The location of the logfile can be changed by using the logfile capability, see
.BR ftpaccess (5).
.LP
The format of the transfer logfile can be customized using the xferlog format
capability, see
.BR ftpaccess (5).
By default, each server entry is composed of a single line of the following
form, with all fields being separated by spaces.
.IP
.I
current-time\ \  transfer-time\ \  remote-host\ \  bytes-transferred\ \  filename\ \  transfer-type\ \  special-action-flag\ \  direction\ \  access-mode\ \  username\ \  service-name\ \  authentication-method\ \  authenticated-user-id\ \  completion-status
.LP
In addition to those in the default format, it also supports chroot-filename,
file-size and restart-offset fields.
.LP
.TP 20
.I current-time
is the current local time in the form "DDD MMM dd hh:mm:ss YYYY". Where DDD
is the day of the week, MMM is the month, dd is the day of the month,
hh is the hour, mm is the minutes, ss is the seconds, and YYYY is the year.
.TP
.I transfer-time
is the total time in seconds for the transfer.
.TP
.I remote-host
is the remote host name.
.TP
.I bytes-transferred
is the number of bytes transferred.
.TP
.I filename
is the absolute pathname of the transferred file.
.TP
.I transfer-type
is a single character indicating the type of transfer. Can be one of:
.RS
.RS
.PD 0
.TP 10
.B a
for an ascii transfer
.TP
.B b
for a binary transfer
.PD
.RE
.RE
.TP
.I special-action-flag
is one or more single character flags indicating any special action taken.
Can be one or more of:
.RS
.RS
.PD 0
.TP 10
.B C
file was compressed
.TP
.B U
file was uncompressed
.TP
.B T
file was tar'ed
.TP
.B _
no action was taken
.PD
.RE
.RE
.TP
.I direction
is the direction of the transfer. Can be one of:
.RS
.RS
.PD 0
.TP 10
.B o
outgoing
.TP
.B i
incoming
.PD
.RE
.RE
.TP
.I access-mode
is the method by which the user is logged in. Can be one of:
.RS
.RS
.PD 0
.TP 10
.B a
for an anonymous user
.TP
.B g
for a passworded guest user (see the
.BR guestgroup
capability in
.BR ftpaccess(5) )
.TP
.B r
for a real, locally authenticated user
.PD
.RE
.RE
.TP
.I username
is the local username, or if anonymous, the ID string given.
.TP
.I service-name
is the name of the service being invoked, usually ftp.
.TP
.I authentication-method
is the method of authentication used. Can be one of:
.RS
.RS
.PD 0
.TP 10
.B 0
none
.TP
.B 1
RFC931 Authentication
.PD
.RE
.RE
.TP
.I authenticated-user-id
is the user id returned by the authentication method.
A * is used if an authenticated user id is not available.
.TP
.I completion-status
is a single character indicating the status of the transfer.  Can be one of:
.RS
.RS
.PD 0
.TP 10
.B c
complete transfer
.TP
.B i
incomplete transfer
.PD
.RE
.RE
.TP
.I chroot-filename
is the pathname of the transferred file relative to the chroot point. This
will differ from the filename field for anonymous and guest users.
.TP
.I file-size
is the size, in bytes, of the file on the server.
.TP
.I restart-offset
is the offset, in bytes, at which the file transfer was restarted (0 when
no restart offset was specified).

.SH FILES
.PD 0
.TP 20
.B @LOGDIR@/xferlog

.SH "SEE ALSO"
.BR ftpd(8),
.BR ftpaccess(5)


CHANGES
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: CHANGES,v 1.47 2016/03/14 23:57:54 wmaton Exp $

###########################################################################

Changes in 2.8.0-CC5: Released

 o  Updates to config.guess and config.sub to modernise the autoconf
    build system.

 o  configure script regenerated.

 o  Describe auto-buf tuning configure option in README.AUTOCONF.

 o  Add option (-R) to disable REST command in cases where download
    managers can abuse the server by opening multiple connections to
    retrieve chunks of the same file.  Based on a patch by Philippe Troin
    (See Debian bug ID 138497).

 o  Some minor man page cleanups.

 o  More syncs to Debian patches dating back to 2.4.2.

 o  Update INSTALL instructions.

 o  Changed the TLS default cipher from ALL:!EXP to exclude most insecure
    or unsupported ciphers within modern versions of OpenSSL.

Changes in 2.8.0-CC4: Released 7 January, 2012

 o  Thanks to a patch by Laurent CONSTANTIN, calling SSL_library_init()
    will no longer create an error within the SSL code related to a lack
    of ciphers.  The patch is applied to src/tlsutil.c.

 o  Another patch by Laurent CONSTANTIN, this time eliminating extra
    carriage returns that aren't necessary when calling syslog().

 o  Added -c option which displays the compile-time options that were used
    to build WU-FTPD. 

Changes in 2.8.0-CC3: Released 6 November, 2011

 o  Many URL updates.

 o  The default email address used by the server in absence of an email
    directive in ftpaccess is ftp-bugs, and is noted in ftpaccess now.

 o  Added code for ADLER-32 checksums for future extension to the list
    of algorithms available to do checksums.

 o  hard-coded links to cksum and md5sum within the FTP root now point to
    these utilities in /bin instead of /usr/bin.

 o  As defined in GFD.20 (GridFTP:  Protocol Extension to FTP for the
    Grid) by the Global Grid Forum, the SBUF command has been introduced.
    This calls the same code as SITE BUFSIZE would.  SITE BUFSIZE is
    still present for backwards compatibility.

 o  Thanks to a contribution by Austin Ellis, added RHEL6 packaging info
    to the codebase in packages/rpm.  This also includes a patch for
    building on that particular system.

 o  Updated ftpd.8 manpage documenting some missing SITE commands that
    had been in the server since 1998.

 o  First cut at implementing the remainder of the RFC3659 extended FTP
    commands.  FEAT, MLST, OPTS and MLSD are now supported.  Based upon
    early gridFTP releases which used WU-FTPD 2.6.0 and 2.6.1 code.

 o  According to draft-somers-ftp-mfxx-04.txt, defined the stubs for the
    MFF, MFCT and MFMT commands.  These may or may not be supported in the
    future, but at the least they are defined for now.  If so, they may
    become SITE commands instead if the spec is not approved.

 o  Updated doc/examples/ftpconversions based on Debian changes introduced
    sometime in 1995.  The server can now deal with such things as bzip2,
    xz, zip and unzip.

 o  If there's an empty /etc/fstab, a segfault can be produced.  A fix
    from Debian BugID 568235 to path_to_device in src/extensions.c addresses
    this.  Patch by Chris Butler.

 o  Added another possible path for quota.h, patch by Chris Butler of
    Debian.

 o  Added missing handler for SIGTERM, patch by Chris Butler of Debian
    related to Debian bug #536774.

 o  Changed -h invocation on the commandline to print out a very brief help
    listing all commandline options that could be supplied to the daemon.
    Also note that some switches are or have an effect on the configuration
    file.  The former function assigned to the 'h' option is now 'H'.

 o  The former function assigned to the 'h' option is now 'H'.  Said it
    twice now.  The manpage has been updated to reflect the changes.

 o  Implemented the CSID command as found in the IETF draft
    draft-peterson-streamlined-ftp-command-extensions.  This is meant
    to be a standards-based replacement for the CLNT command.  Currently
    this command is hard-wired with the server name and version, but the 
    next release will permit customisation.

###########################################################################

Changes in 2.8.0-CC2: Released 26 June, 2009

 o  Cleaned-up and updated the ftpd.8 manpage which had languished for
    some time.

 o  Added Auto tuning/auto buf, with added commands SITE BUFSIZE and
    SITE BUFSIZEMEASURE.  Use --enable-autobuf to enable this feature.
    Based on a patch on work done on WU-FTPD-2.6.1 by Gaurav Navlakha, et al
    of DAST, NLANR (now CAIDA).  This code also contains the ability to
    report back on buffer usage.  Note that an autobuf client needs to
    be used to take advantage of this feature.  For the time being, the
    code for buffer measurement is IPv4 only.

 o  Updated src/extensions.c to take into account a change in Linux quota
    structure.  Based on a patch by Chris Butler.


src/xferlog.c
/****************************************************************************  
 
  Copyright (c) 2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: xferlog.c,v 1.5 2011/10/20 22:58:11 wmaton Exp $
 
****************************************************************************/

#include "config.h"
#include <string.h>
#include "extensions.h"
#include "proto.h"

#define DEFXFERFORMAT	"%T %Xt %R %Xn %XP %Xy %Xf %Xd %Xm %U ftp %Xa %u %Xc"

int xferdone = 0;
struct xferstat xfervalues;
char xferlog_format[MAXXFERSTRLEN] = DEFXFERFORMAT;

/*************************************************************************/
/* FUNCTION  : get_xferlog_format                                        */
/* PURPOSE   : Read the xferlog format string from ftpaccess into        */
/*             xferlog_format if it exists otherwise load default string */
/* ARGUMENTS : none                                                      */
/*************************************************************************/

void get_xferlog_format(void)
{
    int which;
    struct aclmember *entry = (struct aclmember *)NULL;

    /* xferlog format <formatstring> */
    xferlog_format[0] = '\0';
    while (getaclentry("xferlog", &entry)) {
	if (ARG0 && (strcasecmp(ARG0, "format") == 0)) {
	    for (which = 1; (which < MAXARGS) && ARG[which]; which++) {
		if (which > 1) {
		    if (strlcat(xferlog_format, " ",
			sizeof(xferlog_format)) >= sizeof(xferlog_format))
			break;
		}
		if (strlcat(xferlog_format, ARG[which],
		    sizeof(xferlog_format)) >= sizeof(xferlog_format))
		    break;
	    }
	    break;
	}
    }

    /* default xferlog format */
    if (xferlog_format[0] == '\0')
	(void) strlcpy(xferlog_format, DEFXFERFORMAT, sizeof(xferlog_format));
}
